os_net_provider.go

```diff
@@ -0,0 +1,42 @@
+package sockaddr
+
+import (
+	"fmt"
+	"net"
+)
+
+type OSNetProvider struct{}
+
+// GetAllInterfaces iterates over all available network interfaces and finds all
+// available IP addresses on each interface and converts them to
+// sockaddr.IPAddrs, and returning the result as an array of IfAddr.
+func (n *OSNetProvider) GetAllInterfaces() (IfAddrs, error) {
+	ifs, err := net.Interfaces()
```

Comment:
hey @deblasis thanks for following up on the team's latest -- we really appreciate your efforts here and receptiveness to feedback.

I may not have been super clear in a previous comment but the line we want to stub out is this one!

We probably want all of the code here to be executed during the tests, except for this line.

This line should look something like 

```
ifs, err := netProvider.Interfaces()
```

and in the testing we can set up a `FakeNetProvider`. What do you think?

Reply:
Actually, I'm not sure if we can easily inject a `netProvider`-like dependency here since the exported functions are `func(string) (string, error)`.

It will probably be easiest to declare
```go
packages sockaddr

var netInterfaces = net.Interfaces()
```
and stub out `sockaddr.netInterfaces` in tests

Reply:
Hi there, sorry for the delayed reply, I have been very busy freelancing.
I had a quick look and I believe we are oversimplifying. I'll try to explain what I mean:

We can sure stub out `net.Interfaces` but that means that we are stubbing out the hardware information about our interfaces, what about the software part? 
We are forgetting the calls to `intf.Addrs()` that look like they are returning network protocol and IP Addresses for the NICs by querying the RIB (Routing Information Base) which returns the address table.

https://github.com/hashicorp/go-sockaddr/pull/38/files#diff-9753b911177155deeb161b6f775d09bdbb6cb51fb315e9b6e9bc3130affd0b46R21

that internally call `net.interfaceAddrTable`
![image](https://user-images.githubusercontent.com/29378614/150688626-0b96b4e5-5526-4f6d-aec9-3d99d4e5281c.png)

and `net.addrTable`
![image](https://user-images.githubusercontent.com/29378614/150689031-d0acf112-df47-44e8-9508-f52f67fc5ba2.png)

As far as I understand, in order to be able to stub out all the system calls that are involved in computing the return value of our `sockaddr.GetAllInterfaces()`, I think we should also look into stubbing these other system calls as well...

Please correct me if I am wrong and let me know how to proceed.

Reply:
That's a great point @deblasis, I agree that the entire `GetAllInterfaces` function should be mocked.

I think exporting new interfaces and implementation structs to mock it are overkill; could we perhaps simply do
```go
var GetAllInterfaces func() (IfAddrs, error) = getAllInterfaces

func getAllInterfaces() (IfAddrs, error) {
        // implementation ...
}

// ifaddr_test.go
func TestGetInterfaceIP(t *testing.T) {

	orig := sockaddr.GetAllInterfaces
        sockaddr.GetAllInterfaces = mockFunc()
        t.Cleanup(func(){ sockaddr.GetAllInterfaces = orig })

        // tests ...
}

Reply:
Hi @kisunji!

I apologize for the delay.
Sounds good, I have updated the PR accordingly.